lib/src/main.dart

```diff
@@ -64,27 +83,142 @@ class OfflineBuilder extends StatefulWidget {
   /// Used for building the error widget incase of any platform errors
   final WidgetBuilder? errorBuilder;
 
+  /// Maximum number of retry attempts with exponential backoff
+  final int maxRetries;
+
+  /// Minimum duration between manual retry attempts to prevent spam
+  final Duration retryCooldown;
+
+  /// Callback function that gets executed when retry is triggered
+  final RetryCallback? onRetry;
+
+  /// Callback to get access to the OfflineBuilder state for retry functionality
+  final void Function(OfflineBuilderState)? onBuilderReady;
+
   @override
   OfflineBuilderState createState() => OfflineBuilderState();
 }
 
 class OfflineBuilderState extends State<OfflineBuilder> {
   late Stream<List<ConnectivityResult>> _connectivityStream;
 
+  int _retryCount = 0;
+  DateTime? _lastRetryTime;
+  bool _isRetrying = false;
+
   @override
   void initState() {
     super.initState();
 
-    _connectivityStream = Stream.fromFuture(widget.connectivityService.checkConnectivity())
-        .asyncExpand((data) => widget.connectivityService.onConnectivityChanged.transform(startsWith(data)))
-        .transform(debounce(widget.debounceDuration));
+    _connectivityStream =
+        Stream.fromFuture(widget.connectivityService.checkConnectivity())
+            .asyncExpand((data) => widget
+                .connectivityService.onConnectivityChanged
+                .transform(startsWith(data)))
+            .transform(debounce(widget.debounceDuration));
+
+    // Expose state to parent widget for retry functionality
+    widget.onBuilderReady?.call(this);
   }
 
+  @override
+  void dispose() {
+    super.dispose();
+  }
+
+  /// Manually retry connectivity check with exponential backoff
+  Future<void> retry() async {
+    if (_isRetrying) {
+      return;
+    }
+
+    final now = DateTime.now();
+    if (_lastRetryTime != null &&
+        now.difference(_lastRetryTime!) < widget.retryCooldown) {
+      return; // Still in cooldown period
+    }
+
+    if (_retryCount >= widget.maxRetries) {
+      return; // Max retries exceeded
+    }
+
+    setState(() {
+      _isRetrying = true;
+    });
+
+    // Set last retry time at the start
+    _lastRetryTime = now;
+
+    try {
+      // Execute custom retry callback if provided
+      if (widget.onRetry != null) {
+        await widget.onRetry!();
+      }
+
+      // Calculate exponential backoff delay
+      final delaySeconds = (1 << _retryCount); // 2^retryCount
+      await Future.delayed(Duration(seconds: delaySeconds));
+
+      // Force a connectivity check - this will trigger the stream naturally
+      await widget.connectivityService.checkConnectivity();
+
+      _retryCount++;
+    } catch (e) {
+      // Handle retry errors gracefully
+      debugPrint('Retry failed: $e');
```

Comment:
It would be great to communicate this back to the caller instead of logging.

I can imagine that instead of adding another `onRetryFailure` parameter, we can instead introduce some sort of reactive `retryController` that includes both `onRetry` and `onError` that can both be implemented by the caller and also the getters defined here